site/docs/skylark/language.md

```diff
@@ -37,142 +29,97 @@ def fizz_buzz(n):
-some restrictions in order to guarantee thread-safety and determinism. Two
-mutable data structures are available: [lists](lib/list.html) and
-[dicts](lib/dict.html).
+Starlark favors immutability. Two mutable data structures are available: [lists](lib/list.html) and [dicts](lib/dict.html). Changes to mutable data-structures, such as appending a value to a list or deleting an entry in a dictionary are valid only for objects created in the current context. After a context finishes, its values become immutable.
 
-In a build, there are many "evaluation contexts": each `.bzl` file and each
-`BUILD` file is loaded in a different context. Each rule is also analyzed in a
-separate context. We allow side-effects (e.g. appending a value to a list or
-deleting an entry in a dictionary) only on objects created during the current
-evaluation context. Once the code in that context is done executing, all of its
-values are frozen.
+This is because Bazel builds use parallel execution. During a build, each `.bzl` file and each `BUILD` file get their own execution context. Each rule is also analyzed in its own context.
 
-For example, here is the content of the file `foo.bzl`:
+Let's go through an example with the file `foo.bzl`:
 
 ```python
-var = []
+# `foo.bzl`
+var = [] # declare a list
 
-def fct():
-  var.append(5)
+def fct(): # declare a function
+  var.append(5) # append a value to the list
 
-fct()
+fct() # execute the fct function
 ```
 
-The variable `var` is created when `foo.bzl` is loaded. `fct()` is called during
-the same context, so it is safe. At the end of the evaluation, the environment
-contains an entry mapping the identifier `var` to a list `[5]`; this list is
-then frozen.
+Bazel creates `var` when `foo.bzl` loads. `var` is thus part of `foo.bzl`'s context. When `fct()` runs, it does so within the context of `foo.bzl`. After evaluation for `foo.bzl` completes, the environment contains an immutable entry, `var`, with the value `[5]`.
 
-It is possible for multiple other files to load symbols from `foo.bzl` at the
-same time. For this reason, the following code is not legal:
+When another, `bar.bzl` loads symbols from `foo.bzl`, loaded values remain immutable. For this reason, the following code in `bar.bzl` is illegal:
 
 ```python
-load(":foo.bzl", "var", "fct")
+# `bar.bzl`
+load(":foo.bzl", "var", "fct") # loads `var`, and `fct` from `./foo.bzl`
 
 var.append(6)  # runtime error, the list stored in var is frozen
 
 fct()          # runtime error, fct() attempts to modify a frozen list
 ```
 
-Evaluation contexts are also created for the analysis of each custom rule. This
-means that any values that are returned from the rule's analysis are frozen.
-Note that by the time a custom rule's analysis begins, the .bzl file in which
-it is defined has already been loaded, and so the global variables are already
-frozen.
-
+Just like the above example using `bzl` files, values returned by rules are immutable too.
```

Comment:
I'd preserve the caveat that rules can't mutate globals in the same file. Maybe add the sentence:

"Since a rule function runs after the .bzl file it's in has already been loaded, rules can't update global variables."

Reply:
Thank you! I added the sentence.